### PR TITLE
fix: remove mandatory index signature

### PR DESCRIPTION
**Pull request by aigoncharov**
Currently, `type State` in `zustand` forces all users to add an index signature to their state interfaces.

```ts
// We have to add 'extends Record<string, any>'
// which adds an index signature that we do not really want
export interface JediStore extends Record<string, any> {
  jediName: string
  jediPower: number
}
export const useChartComposerDataStore = create<JediStore>(...)
```

With this fix the index signature is no longer mandatory.

### CODE REVIEW

File: src/vanilla.ts
Comment (dai-shi) [inline src/vanilla.ts:50]: I didn't know the 37663 issue, but hit the issue behavior in other projects. If this is only about typing, can we just add a type assertion until the issue is resolved? That's what I do for all other projects.

**Review by dai-shi — COMMENTED** [review]
This is really nice. Really wanted someone with better TS knowledge to come.

File: src/vanilla.ts
Comment (aigoncharov) [reply src/vanilla.ts]: I was torn between type assertion and `instanceof` as well. Seems like tomayto/tomahto kind of thing to me. Fixed in https://github.com/pmndrs/zustand/pull/376/commits/9a48fa8dce62fadf75866a4a0569db857e7ab9f0

File: src/vanilla.ts
Comment (aigoncharov) [reply src/vanilla.ts:5]: Thank you for creating `zustand` in the first place! Updated the TS-related section - https://github.com/pmndrs/zustand/pull/376/commits/a40c32c261d875014c4d3833cfd726e2dc92ce8f

Comment (aigoncharov) [reply src/vanilla.ts:5]: @dai-shi regarding `object`, I think this one should more or less cover it https://stackoverflow.com/a/61000565/10687034
I created a small example 
https://www.typescriptlang.org/play?strict=true#code/C4TwDgpgBAkgygVwEYBVwQPIDMCiBbMUDJAKwB4UA+KAXihSggA9gIA7AEwGcoBvAXygB+KMABOCaAC4oWAIYAbLhABQoSLESp02fIRAAlCAGMA9mI4VqdBs1aceRsxbJdxASzYBzKAB8obAh4SBBiflBcIMGmCgA0UHJsINQi4pJQMvJKqmroUADCABYmANYAjHK0mshokLoERKRkvEhyYjIATPzUAPQ9UCAQPO7AUO5cuRpFpR2VdPA1OrgNIMTkvFimpp3xre1QAORyB91QfQNDY6Pjk9DTxiUAzHPV2nXL+mtkgcGhvf1cBBiMBicaeLwKECiQpya48RRcUyiCQ5W4FYoPMpIKoLN6YD6gJzmSwtNqdU7nQbDOFo+4lDrY+ZaWr4vSEkzE5qbbZQDq7MmHY4U-pUq5jCbqO4Yp6M14s+r6IkuH4hMT-eiFcZQUxsaBarLKBI8ZiQYysDhAA

Comment (aigoncharov) [reply src/vanilla.ts:5]: I have double-checked `combine`. It seems to keep working with the updated typings https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcDKMCGNUF44BKwAxtACYA8AzjFAJYB2A5nAD5wMCuAtgEbBQ2cKoj4QANgBo4mBogB8AKEWhIsBMlQAFTLDqZxGbMAoAVOKBwMyVdFhzSA0hZBWbcANbBEEAGZxzfC8ff1N5OFxFOCEACi06Yg8zJ3D2UwBKKNiYmmMALgD0iPD4xOS4R1TC5RVwaHgkFDtjNGBxEhhoMxc3WyMHOABVcPwc+2ACjOKh2rUGzTgAUQBHTgM6JABhAAsSLygzEbgx-IDpBmAAd36JwuneCAlgWVn6jSabgBk6GmALg7CEWOuRwk2kYCgwAAbnQIJwqDdJkVcOEoRA6GRXupGqgbmhxAlgN9fv9DkCcgTiLdTODITC4QjKdTkaj0ZjVG9GDgoL5MFT0JxeFRiPR+N1LH93DdwgBvLIxAkkgQFL4-KwCQ7pAoxFlwNEYrIUPFM+QxLLRRXqqAq8b4wnEq1G20myTm4RtDrQG0tD3EToA6TGwnyV3RaLAVbrJAAMQYAH4Cis1gStrtEhqg1SlNEtcddfrMQBfGoc7ELVowG7i1ySvrjI5y6IUZwS6y2YJ+AKmsC6GD6cQFHR6AxVmkVENwSFgcR84AJuAPJ6yXMFxTF0vzJoAcWAlfG1d6zRwRx101Myg371QABFgDQoBBEOT82ysfAuQJefyMNBgABBMA6APWsj2AWUsioXdEXQKD9zCLImFg04dz3YxDggwVhVFW40EwkU6DFeDojIO9aEfApb3vR81xLOoyw+cZNkhbAunMVspXGaRNnhTpuArG4gX4uD5BPCDdwKbiaAgPikIcBDxLgFDRxDLJMEAm1fwAoD4N1c83zgUgGBoAzpN4Rg8DgCgsi0ehuF0RABPYutjFDGDDLIezHJrNtQMUU0skYDZ+2gmy6DsqAHM4rIRWeUFjjdSCYBVWSTFC8LIuMFSwzgRCksUlKKDSzz61c6I1LoDTIS0wrbOKzKsl1Vp3LqnBFFzG4mNiroioigSADI3IgawWrA6Ycl3aRcukcqWSyAB5XgACsOgAOkwKgqDoJgGDNbKZULUq4ECvsRyi7KYuMXbsvdeB1pg1CcBqsKRqy67cpkWwlP3HqMuPQ6ysAj67E0wCnvS6U3UyHNlGiZRDOM1QOlMBZ8GSU0oQMThmQKAtplxgAGABuOGhuM+FgB-SEgWi0zzKu6IZQXZ4oCoAp8bgA64DdcaYF1GJGcYC7IO1XhEAKLg+AEXVEpiE4cD5xn+F0VnhHGFalZZuAAGoFyfQt0iKfWGovcAkc0J1jE6liDkV5mVYl-goEJo6GCF24YlF8WeEdl8MQ5kSYnJyngEyZQgA

**Comment by dai-shi** [comment]
https://github.com/pmndrs/zustand/pull/376#discussion_r625737319
Thanks for the investigation. You compare `{}` and `Record<string | number | symbol, any>`, but what I mean is to compare with `object`.
https://www.typescriptlang.org/play?ssl=2&ssc=43&pln=2&pc=80#code/C4TwDgpgBAkgygVwEYBVwQPIDMCiBbMUDJAKwB4UA+KAXihSggA9gIA7AEwGcoBvAXygB+KMABOCaAC4oWAIYAbLhABQoSLESp02fIRAAlCAGMA9mI4VqdBs1aceRsxbJdxASzYBzKAB8obAh4SBBiflBcIMGmCgA0UHJsINQi4pJQMvJKqurQ8MhokNjEJCbAVrT0jCzs3FCmpGXCohLSsorKKmroUADCABYmANYAjHKV+dpFuAREpGS8SHJiMgBM-NQA9JtQIBA87sBQ7lzdGgPDq+N0k4WYM-olC1impmvxSytQAORy3xtQba7fbHI4nM7QC7GIYAZmumgKOgec3IgWCoS2Oy4CDEYDEJ08XgUIFE-TkYJ4HVMLUkXQhfUG0JGSAmWjuulmhhM5ksi2WawBQL2Bwp9KhQ1WLJubKRelATh5z1e7ygnxkv3+mOBIuOp1yDOGMKlCKm9zlXOcljRITEWpQ-RO9TY0EdWWUCR4zEgxlYHC6+vFI2MrMR0xKZQWaqg6y1wtBurFjIlwelofu4Z9Sre0Y+-J+f0FOzjhwTAaTMJTJvZWAz5WtGMBO3tjtMzt17WyHuq3t9oleKiAA

If there's zero difference, I might prefer `type State = object`.

File: src/vanilla.ts
Comment (dai-shi) [reply src/vanilla.ts]: Thanks! It will keep the bundle JS code identical.

File: src/vanilla.ts
Comment (dai-shi) [reply src/vanilla.ts:5]: Thanks for updating readme.
As for `Record` vs `object`, I replied in the main thread. Making this conversation resolved.

**Comment by dai-shi** [comment]
Reading: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md

> - Avoid the `object` type, as it is currently hard to use due to not being able to assert that keys exist.
>   - See https://github.com/microsoft/TypeScript/issues/21732.

Hmm, I'm not sure what "assert that keys exist" means.

---

At least the original issue is resolved.
https://www.typescriptlang.org/play?#code/C4TwDgpgBAysCGwIFUB2BrVB7A7qqAvFAEoQDGWATgCYA8AzsJQJaoDmUAPlKgK4C2AIwiUuUeiCFYANgBoovDNjwA+AFChIsBEgCCqEIRLkqdRi3Zi+QkWIlS5UeAfWbocRBADyggFblgIyw-ALU1ClRGKAAzVDRMXHwiWgAVKAgADyRUanptT3jlVBUACnhKNgAuKBSASkIVJwrwrEjA2P1DZLTM7Nz8vRcyiuq6hqa2FraY1B9-MkDu9KyIHLyPJDmA0vKqmvqCRt2w1iRKaPgyaAApCGpmDegAbzUoN6h-e4A5eH4IavMrEm7w+d2YAAVcCJqtZhJQ1ABfKZREBYajwaq3e6PIwvEGfZg-P7VADkAE00fASbJXu8CZCcNCoABGAAMiLCsUKiRKqPRtTUHQMvMpAtiWwWIv5QA

**Comment by aigoncharov** [comment]
> [#376 (comment)](https://github.com/pmndrs/zustand/pull/376#discussion_r625737319)
> Thanks for the investigation. You compare `{}` and `Record<string | number | symbol, any>`, but what I mean is to compare with `object`.
> https://www.typescriptlang.org/play?ssl=2&ssc=43&pln=2&pc=80#code/C4TwDgpgBAkgygVwEYBVwQPIDMCiBbMUDJAKwB4UA+KAXihSggA9gIA7AEwGcoBvAXygB+KMABOCaAC4oWAIYAbLhABQoSLESp02fIRAAlCAGMA9mI4VqdBs1aceRsxbJdxASzYBzKAB8obAh4SBBiflBcIMGmCgA0UHJsINQi4pJQMvJKqurQ8MhokNjEJCbAVrT0jCzs3FCmpGXCohLSsorKKmroUADCABYmANYAjHKV+dpFuAREpGS8SHJiMgBM-NQA9JtQIBA87sBQ7lzdGgPDq+N0k4WYM-olC1impmvxSytQAORy3xtQba7fbHI4nM7QC7GIYAZmumgKOgec3IgWCoS2Oy4CDEYDEJ08XgUIFE-TkYJ4HVMLUkXQhfUG0JGSAmWjuulmhhM5ksi2WawBQL2Bwp9KhQ1WLJubKRelATh5z1e7ygnxkv3+mOBIuOp1yDOGMKlCKm9zlXOcljRITEWpQ-RO9TY0EdWWUCR4zEgxlYHC6+vFI2MrMR0xKZQWaqg6y1wtBurFjIlwelofu4Z9Sre0Y+-J+f0FOzjhwTAaTMJTJvZWAz5WtGMBO3tjtMzt17WyHuq3t9oleKiAA
> 
> If there's zero difference, I might prefer `type State = object`.

For some reason, I thought `{}` and `object` are aliases to the same thing. You live - you learn :)
I haven't found any evidence that `Record<string | number | symbol, any>` should be preferred over `object`. However, they behave a bit differently. With `object` we'll need a type assertion for `combine`.
https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcDKMCGNUF44QBGAVsAMYwBQloksCyqACprAJaYA2G2wAPABU4oHADsAJgGd0WHABo4AaWEgxUuAGtgiCADM4Q-Fp36BAPji5KcOAB84ACiZsyGwQsUX7AgJTW7jg6SssAAXAY+lhbOru5KXhHUNODQ8EgoMrxowJzkMNCCKmrSPPJwAKoW+EEh4b5RFcl0aYxwAKIAjgCuXGxIAMIAFuRaUIJVjsG8dQqiwADupWERDYQQELmYok2pDBlLADJswcBzY+aWk7UGCmBQwABubBBdkkt1kbgWDxBs4jv0dKoJZoTguYBHE5ncaXIJgsjLAS3e5PF5veGIz7fX7-Wi7NiiHBQXSYBHoLqESRkKBsQj8IQiU7qJYWADe-gcYKhwCg4UOxzEPPGPnCDixcB+f38fBBGLMDn8Ni5gt5mRwoPBkJVMpCGoRZjkirgkhyeWgfN1pooBSRauAeuABqNwG6vSQADFRAB+cKdHpggbDVxC2Xgsz+EWOcWS-4AXySeMBrWyMCWhUZEhKIQm7JsfGUGfUxj0BnlYFYMA4nHCLHYXDTts8CnuYE4pOAPrgaw2wC2kZjlHjiZaGQA4sBUyF06omVneBMxQ0BNRh3tUAARYDBKAQRCw6M4gHwAlEklkjDQYAAQTAbGnxTtbP8Jsn03QE4b4ZsAHMP9dx6+ODjM+FJUjSdJ8qB1K0vSX5wOIW4wDuiDhJu267oOCYpEm+whP09zYDaRSznaCj9K8+QALYpkslw0VO5gLs+E7hORwQQNRf68IaP4sXAAGfjxcCYLeFqXjed7mJQ4rLkecBkBAojBPJHGEASeBwHw-hMDSlGsIgtGFnO8j+NkCkSPphkzpmdqUPK-gEn0VbvHAOlsHpUAGSEQnUr2OCikaL58lxQFuR5XnzkJvEwOEAlTmFlnZlFwmiTI4m3nwCWeSyEYNGZiniIlvDSRavD4X5BRZRFOBwAAZO+5mFdl2YNEEE4KL+MAKCJbBYv4ADyJB5AAdJgkiSGw36iAqNg2KysbJY5lb1t5Rq+bwM2zTYL7CSUIX8FVLLJdFu38ftmW6UVOBOltNg9adF73BJF3uVdjpGn4Nh+P41DmcptB5AIrT4O48oPFwXSYuEMYNDDAAMADcv2Kcprz2vk9yXP4CmUWpcybXNXa9lAkjhHDcALXARptTA4oOKycAEutJqioQKFwKIXS4zy4ovg4NS8HTDN0qwpPGiEw0iyTcAANRdnusY+JEisRiu4CA4wOplQRGN8MLxNi5z3NQAjjOiMzywOGz4RG3SUAHn8FNmPKaOPcA32UEAA

**Comment by dai-shi** [comment]
Thanks! I'd prefer `object` for no specific reason, if there were no difference. Until we learn something new, let's go with `Record`.

**Review by dai-shi — APPROVED** [review]

**Comment by dai-shi** [comment]
Actually, my ideal goal is to make it `State = unknown` and disallow any object, but still want to ensure state to be object (`number` is forbidden, for example).

**Comment by dai-shi** [comment]
2a143c6 👁️

**Comment by aigoncharov** [comment]
@dai-shi I would prefer `object` as well, just for simplicity. 
As we can see from this commit, `object` is still a bit different -https://github.com/pmndrs/zustand/pull/376/commits/2a143c6bdfdb2a352c44a5055717b8118da05312. Not in a bad way necessarily.
BTW, was there a specific reason to use `extends any` for `shallow`? My understanding that everything extends any by default. I tried to remove it in https://github.com/pmndrs/zustand/pull/376/commits/2a143c6bdfdb2a352c44a5055717b8118da05312 and it simplified the typings in `shallow` a bit.
Let me know if we need to revert https://github.com/pmndrs/zustand/pull/376/commits/2a143c6bdfdb2a352c44a5055717b8118da05312

File: src/middleware.ts
Comment (dai-shi) [inline src/middleware.ts]: Just curious why this change is required.

File: src/middleware.ts
Comment (dai-shi) [inline src/middleware.ts:139]: This looks unfortunate, but can be accepted.

File: src/middleware.ts
Comment (dai-shi) [inline src/middleware.ts:244]: I would do this. Does it work?
```ts
      ;(Object.keys(state) as (keyof S)[]).forEach((key) => {
        !whitelist.includes(key) && delete state[key]
```

File: src/shallow.ts
Comment (dai-shi) [inline src/shallow.ts:1]: This looks good. Thanks for the fix. (It was more or less a historical reason.)

File: tests/basic.test.tsx
Comment (dai-shi) [inline tests/basic.test.tsx:680]: I see. Looks more precise.
BTW, we should remove `extends State` in L664.

**Review by dai-shi — COMMENTED** [review]
This looks good.

File: src/middleware.ts
Comment (dai-shi) [reply src/middleware.ts]: already resolved.

File: src/middleware.ts
Comment (aigoncharov) [reply src/middleware.ts]: Addressed in https://github.com/pmndrs/zustand/pull/376/commits/8473244cce1fa4a10d8a63cc977bae50508d0813

File: src/middleware.ts
Comment (aigoncharov) [reply src/middleware.ts:139]: Agreed. I haven't found a way around it :(

File: tests/basic.test.tsx
Comment (aigoncharov) [reply tests/basic.test.tsx:680]: I wish we didn't have to do it, but without the explicit second argument it does not pass the assertion in `checkAllTypes` after our migration to `object`. I would say it even makes sense :)
At the same time, if we need to make all these changes in the library code because of the migration to `object`, there is a risk it might break code that consumes the library.
I see two ways to go from here to mitigate the risk:
1. Migrate to `object` and make it a minor release, not a patch one.
2. Revert changes related to `object` migration and keep using `Record`.

Personally, I would suggest option #1.

File: src/middleware.ts
Comment (aigoncharov) [reply src/middleware.ts]: FYI, it was leftover from my experiment with removing the default value for the second generic.

File: tests/basic.test.tsx
Comment (dai-shi) [reply tests/basic.test.tsx:680]: Yes, this looks good and more like what I wanted. Let's go with minor, and describe breaking changes in types in the release note.

**Review by dai-shi — APPROVED** [review]

**Comment by aigoncharov** [comment]
@dai-shi thank for the conversation and for your review! What is the process after your approval?

**Comment by dai-shi** [comment]
It depends. Sometimes, we just wait for others to review.
For this one, I will merge it and make a release soon.

**Comment by dwelle** [comment]
> Avoid the object type, as it is currently hard to use due to not being able to assert that keys exist.
See microsoft/TypeScript#21732.

Just came upon this topic. FYI, what the above refers to is the fact that `object` has no index signature (unlike `Record`), and you cannot use the `in` as a type guard ([playground](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=13&pc=2#code/MYewdgzgLgBCMF4YG8C+MCGE4CMBWApsFANwBQZAlgGYwAUA5BgzJWHAJQpkwyiQgANgQB0gkAHM6IERg7lUFAPRKYAWg2at2nbr1aK-aHABMiFOiwoMALhgAWE+gA+KHHYCMABi+pyVWkZmVnYQEy5kHj5wCCFRcSkw2XkyVCA)):

```
if ('a' in o) {
	return o.a; // does not type-check for `object` type
}
```

**Comment by dai-shi** [comment]
Thanks. Okay, that's actually what I probably [wanted](https://github.com/pmndrs/zustand/pull/376#issuecomment-832477249). So, this should be a valid use case of `object`.

[playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBCMF4YG8C+MCGE4CMBWApsFANwBQAlgGYwAUA5BvTBWHAJQpkwyiQgAbAgDoBIAOa0QwjO3KoyZPtDgAmRCnRYUGAPwAuGABZVqctTqNmrNZ2Tde4CIJFjJIVTLlkFSp7BAAZg00TGwAJSIQACcAEwAeaGjWcQAaGABXMABrMBAAdzAAPkoaBiYWNiC7B2UXUQkpQK95RWUAoxCtCKi4xKhksDTMMABPEoty6yqjGp46oQb3IxafMiA)

**Pull request merged**
